### PR TITLE
First pass to create a working hyper+soketto example

### PR DESCRIPTION
**Pull request by jsdw**
Addresses #44.

This is a working Soketto+Hyper example. We may not wish to merge this as-is though; it may be better to use this as inspiration for which parts could be exposed/added to Soketto (possibly behind a "hyper" feature flag) to make this simpler (and more solid!). (alternately, we could do what https://github.com/de-vri-es/hyper-tungstenite-rs did (which was a source of inspiration for this code) and release a separate crate).

### Testing

Start the server with:

```
cargo run --example hyper_server
```

Navigate to http://localhost:3000.

In a browser console, run:

```
var socket = new WebSocket("ws://localhost:3000");
socket.onmessage = function(msg) { console.log(msg) };
socket.send("Hello!!");
```

Any messages sent should be echoed back.

### CODE REVIEW

File: examples/hyper_server.rs [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+// To try this, start up the example (`cargo run --example hyper_server`) and then
+// navigate to localhost:3000 and, in the browser JS console, run:
+//
+// ```
+// var socket = new WebSocket("ws://localhost:3000");
+// socket.onmessage = function(msg) { console.log(msg) };
+// socket.send("Hello!");
+// ```
+//
+// You'll see any messages you send achoed back.
```
Comment (insipx) [inline examples/hyper_server.rs]:
```suggestion
// You'll see any messages you send echoed back.
```

**Review by insipx — APPROVED** [review]
It works when I ran it and the code looks good :+1:

File: examples/hyper_server.rs (hunk 1 above)
Comment (jsdw) [reply examples/hyper_server.rs]: Thanks!

File: examples/hyper_server.rs [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+/// Return the response to the upgrade request, and a way to get hold of the underlying TCP stream
+fn upgrade_to_websocket(req: Request<Body>) -> Result<(Response<Body>, hyper::upgrade::OnUpgrade), handshake::Error> {
+	let key = req.headers().get("Sec-WebSocket-Key").ok_or(handshake::Error::InvalidSecWebSocketAccept)?;
+	if req.headers().get("Sec-WebSocket-Version").map(|v| v.as_bytes()) != Some(b"13") {
+		return Err(handshake::Error::HeaderNotFound("Sec-WebSocket-Version".into()));
+	}
+
+	// Just a little ceremony we need to go to to return the correct response key:
```
Comment (jsdw) [inline examples/hyper_server.rs]:
```suggestion
	// Just a little ceremony we need to go through to return the correct response key:
```

File: examples/hyper_server.rs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+	header_contains_value(request.headers(), hyper::header::CONNECTION, b"upgrade")
+		&& header_contains_value(request.headers(), hyper::header::UPGRADE, b"websocket")
+}
+
+/// Check if there is a header of the given name containing the wanted value.
+fn header_contains_value(headers: &hyper::HeaderMap, header: hyper::header::HeaderName, value: &[u8]) -> bool {
+	pub fn trim(x: &[u8]) -> &[u8] {
+		let from = match x.iter().position(|x| !x.is_ascii_whitespace()) {
+			Some(i) => i,
+			None => return &x[0..0],
```
Comment (niklasad1) [inline examples/hyper_server.rs]: I suppose you could just return an empty slice for readability....

**Review by niklasad1 — APPROVED** [review]

File: examples/hyper_server.rs (hunk 3 above)
Comment (jsdw) [reply examples/hyper_server.rs]: That's a very good point! I probably assumed I needed to preserve the lifetime when I ported this code over, but of course it'll happily be static :)

**Comment by jsdw** [comment]
Unless there are any objections, I'll merge this PR, and when I (or somebody else) gets the chance to make working with hyper a little more ergonomic we can just update this example to match.

**Pull request merged**